include/crm/common/xml.h

```diff
@@ -312,6 +312,48 @@ int xml_apply_patchset(xmlNode *xml, xmlNode *patchset, bool check_version);
 
 void patchset_process_digest(xmlNode *patch, xmlNode *source, xmlNode *target, bool with_digest);
 
+/* exclusive (should any combination make sense -> explicitly enumerated),
+   with intentional symbolic constant duality (easier to adapt to growth) */
+enum pcmk_acl_cred_type {
+#define PCMK_ACL_CRED_USER PCMK_ACL_CRED_USER
+    PCMK_ACL_CRED_USER = 1,
+    /* XXX no proper support for groups yet */
+};
+
+/* need to be ORable */
+enum pcmk_acl_verdict {
+    PCMK_ACL_VERDICT_WRITABLE = 1 << 0,
+    PCMK_ACL_VERDICT_READABLE = 1 << 1,
+    PCMK_ACL_VERDICT_DENIED   = 1 << 2,
+};
+
+/*!
+ * \brief Mark CIB with namespace-encoded result of ACLs eval'd per credential
+ *
+ * \param[in] cred_type        credential type that \p cred represents
+ * \param[in] cred             credential whose ACL perspective to switch to
+ * \param[in] cib_doc          XML document representing CIB
+ * \param[out] acl_evaled_doc  XML document representing CIB, with said
+ *                             namespace-based annotations throughout
+ *
+ * \return  0 if ACLs were not applicable, >0 if it was and all went fine
+ *          (this is the only case when it's safe to touch \p acl_evaled_doc
+ *          afterwards, the result is #PCMK_ACL_VERDICT_WRITABLE,
+ *          #PCMK_ACL_VERDICT_READABLE and #PCMK_ACL_VERDICT_DENIED bits
+ *          ORed respectively), -2 on run-time unrecognized \p cred_type,
+ *          -3 on unsupported validation schema version (see below),
+ *          or -1 on any other/generic issue
+ *
+ * \note Only supported schemas are those following acls-2.0.rng, that is,
+ *       those validated with pacemaker-2.0.rng and newer (artificially caped
+ *       at 4.0, not including it (the future cannot be predicted,
+ *       compatibility needs to be confirmed, then, hopefully leading to
+ *       this comment being updated).
+ */
+int pcmk_acl_evaled_as_namespaces(enum pcmk_acl_cred_type cred_type,
+                                  const char *cred, xmlDoc *cib_doc,
+                                  xmlDoc **acl_evaled_doc);
+
```

Comment:
Let's make all of this internal API, at least for now. I'd rather keep the ability to change it freely until we've had some experience with it.